packages/http/libraries/fastify/src/adapter/InversifyFastifyHttpAdapter.ts

```diff
@@ -0,0 +1,219 @@
+      | undefined,
+  ): void {
+    const router: FastifyPluginCallback = (
+      fastifyInstance: FastifyInstance,
+      _opts: Record<string, unknown>,
+      done: () => void,
+    ) => {
+      const orderedMiddlewareList:
+        | RequestHandler<FastifyRequest, FastifyReply, (err?: Error) => void>[]
+        | undefined = this.#orderMiddlewares(guardList, middlewareList);
+
+      const fastifySyncMiddlewareList: preHandlerHookHandler[] =
+        this.#buildFastifySyncMiddlewares(orderedMiddlewareList ?? []);
+
+      for (const middleware of fastifySyncMiddlewareList) {
+        fastifyInstance.addHook('preHandler', middleware);
+      }
+
+      for (const routerParam of routerParams) {
+        const orderedMiddlewareList:
+          | RequestHandler<
+              FastifyRequest,
+              FastifyReply,
+              (err?: Error) => void
+            >[]
+          | undefined = this.#orderMiddlewares(
+          routerParam.guardList,
+          routerParam.middlewareList,
+        );
+
+        const fastifySyncMiddlewareList: preHandlerHookHandler[] =
+          this.#buildFastifySyncMiddlewares(orderedMiddlewareList ?? []);
+
+        fastifyInstance.route({
+          handler: routerParam.handler as RouteHandlerMethod,
+          method: routerParam.requestMethodType,
+          preHandler: fastifySyncMiddlewareList,
+          url: routerParam.path,
+        });
+      }
+
+      done();
+    };
+
+    this.#app.register(router, { prefix: path });
+  }
+
+  #orderMiddlewares(
+    guardList:
+      | RequestHandler<FastifyRequest, FastifyReply, (err?: Error) => void>[]
+      | undefined,
+    middlewareList:
+      | RequestHandler<FastifyRequest, FastifyReply, (err?: Error) => void>[]
+      | undefined,
+  ):
+    | RequestHandler<FastifyRequest, FastifyReply, (err?: Error) => void>[]
+    | undefined {
+    let orderedMiddlewareList:
+      | RequestHandler<FastifyRequest, FastifyReply, (err?: Error) => void>[]
+      | undefined = undefined;
+
+    if (guardList !== undefined || middlewareList !== undefined) {
+      orderedMiddlewareList = [];
+
+      if (guardList !== undefined) {
+        orderedMiddlewareList.push(...guardList);
+      }
+
+      if (middlewareList !== undefined) {
+        orderedMiddlewareList.push(...middlewareList);
+      }
+    }
+
+    return orderedMiddlewareList;
+  }
+
+  #buildFastifySyncMiddlewares(
+    middlewareList: RequestHandler<
+      FastifyRequest,
+      FastifyReply,
+      (err?: Error) => void
+    >[],
+  ): preHandlerHookHandler[] {
+    return middlewareList.map(
+      (
+        middleware: RequestHandler<
+          FastifyRequest,
+          FastifyReply,
+          (err?: Error) => void
+        >,
+      ) =>
+        (
+          request: FastifyRequest,
+          reply: FastifyReply,
+          done: (err?: Error) => void,
+        ) => {
+          void Promise.resolve(middleware(request, reply, done));
+        },
+    );
+  }
```

Comment:
```suggestion
  #buildFastifySyncMiddlewares(
    middlewareList: RequestHandler<
      FastifyRequest,
      FastifyReply,
      (err?: unknown) => void
    >[],
  ): preHandlerAsyncHookHandler[] {
    return middlewareList.map(
      (
        middleware: RequestHandler<
          FastifyRequest,
          FastifyReply,
          (err?: unknown) => void
        >,
      ) =>
        async (request: FastifyRequest, reply: FastifyReply) => {
          return new Promise<void>(
            (resolve: () => void, reject: (reason?: unknown) => void) => {
              const done: (err?: unknown) => void = (err?: unknown) => {
                if (err !== undefined) {
                  // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
                  reject(err);
                } else {
                  resolve();
                }
              };

              void Promise.resolve(middleware(request, reply, done));
            },
          );
        },
    );
  }
```

Ok, I see `Error` errors are required in the sync hooks interface. Let's use async hooks to properly throw non `Error` values